celery/contrib/testing/mocks.py

```diff
@@ -13,6 +13,13 @@
         from mock import Mock
 
 
+class PicklableMock(Mock):
```

Comment:
@johnarnold would you mind reminding me why this change is required? I also noticed that the `Mock` instance is losing its arguments, so is not fully restored. I can see that `Mock` instances are not pickleable in general, but I was wondering why this is now necessary.

I searched through our previous discussions and I don't think I found any reference. If there is any, please point me to it. Thanks!

Reply:
Sure, without it you get serialization errors.  It's a shame that TravisCI truncated those builds already, I would just link you.

There are many of them, but here is an example:

```
self = <t.unit.worker.test_request.test_Request object at 0x7fbc7a1edc18>

    def test_on_failure_Terminated(self):
        einfo = None
        try:
            raise Terminated('9')
        except Terminated:
            einfo = ExceptionInfo()
        assert einfo is not None
        req = self.get_request(self.add.s(2, 2))
>       req.on_failure(einfo)

worker/test_request.py:264: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../celery/worker/request.py:363: in on_failure
    'terminated', True, string(exc), False)
../../celery/worker/request.py:263: in _announce_revoked
    self.id, reason, request=self, store_result=self.store_errors,
../../celery/backends/base.py:191: in mark_as_revoked
    traceback=None, request=request)
../../celery/backends/base.py:322: in store_result
    request=request, **kwargs)
../../celery/backends/base.py:694: in _store_result
    self.set(self.get_key_for_task(task_id), self.encode(meta))
../../celery/backends/base.py:272: in encode
    _, _, payload = self._encode(data)
../../celery/backends/base.py:276: in _encode
    return dumps(data, serializer=self.serializer)
../../../../venv/celery/lib/python3.6/site-packages/kombu/serialization.py:221: in dumps
    payload = encoder(data)
/usr/lib/python3.6/contextlib.py:99: in __exit__
    self.gen.throw(type, value, traceback)
../../../../venv/celery/lib/python3.6/site-packages/kombu/serialization.py:54: in _reraise_errors
    reraise(wrapper, wrapper(exc), sys.exc_info()[2])
../../../../venv/celery/lib/python3.6/site-packages/vine/five.py:178: in reraise
    raise value.with_traceback(tb)
../../../../venv/celery/lib/python3.6/site-packages/kombu/serialization.py:50: in _reraise_errors
    yield
../../../../venv/celery/lib/python3.6/site-packages/kombu/serialization.py:221: in dumps
    payload = encoder(data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = {'args': None, 'children': [], 'date_done': datetime.datetime(2018, 8, 15, 16, 37, 26, 979482), 'kwargs': None, ...}
dumper = <built-in function dumps>

    def pickle_dumps(obj, dumper=pickle.dumps):
>       return dumper(obj, protocol=pickle_protocol)
E       kombu.exceptions.EncodeError: Can't pickle <class 'unittest.mock.Mock'>: it's not the same object as unittest.mock.Mock

../../../../venv/celery/lib/python3.6/site-packages/kombu/serialization.py:350: EncodeError

Reply:
Sure, there is a serialization error on Mock in Kombu when running t/unit/worker/test_request.py.
Here's a chunk of a bunch of tracebacks:
```
During handling of the above exception, another exception occurred:

self = <t.unit.worker.test_request.test_Request object at 0x7fbc79ff8630>

    def test_on_failure_WorkerLostError_redelivered_None(self):
        einfo = None
        try:
            raise WorkerLostError()
        except:
            einfo = ExceptionInfo(internal=True)
        req = self.get_request(self.add.s(2, 2))
        req.task.acks_late = True
        req.task.reject_on_worker_lost = True
        req.delivery_info['redelivered'] = None
>       req.on_failure(einfo)

worker/test_request.py:342: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../celery/worker/request.py:367: in on_failure
    self.id, exc, request=self, store_result=self.store_errors,
../../celery/backends/base.py:157: in mark_as_failure
    traceback=traceback, request=request)
../../celery/backends/base.py:322: in store_result
    request=request, **kwargs)
../../celery/backends/base.py:694: in _store_result
    self.set(self.get_key_for_task(task_id), self.encode(meta))
../../celery/backends/base.py:272: in encode
    _, _, payload = self._encode(data)
../../celery/backends/base.py:276: in _encode
    return dumps(data, serializer=self.serializer)
../../../../venv/celery/lib/python3.6/site-packages/kombu/serialization.py:221: in dumps
    payload = encoder(data)
/usr/lib/python3.6/contextlib.py:99: in __exit__
    self.gen.throw(type, value, traceback)
../../../../venv/celery/lib/python3.6/site-packages/kombu/serialization.py:54: in _reraise_errors
    reraise(wrapper, wrapper(exc), sys.exc_info()[2])
../../../../venv/celery/lib/python3.6/site-packages/vine/five.py:178: in reraise
    raise value.with_traceback(tb)
../../../../venv/celery/lib/python3.6/site-packages/kombu/serialization.py:50: in _reraise_errors
    yield
../../../../venv/celery/lib/python3.6/site-packages/kombu/serialization.py:221: in dumps
    payload = encoder(data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = {'args': None, 'children': [], 'date_done': datetime.datetime(2018, 8, 15, 16, 37, 27, 460001), 'kwargs': None, ...}
dumper = <built-in function dumps>

    def pickle_dumps(obj, dumper=pickle.dumps):
>       return dumper(obj, protocol=pickle_protocol)
E       kombu.exceptions.EncodeError: Can't pickle <class 'unittest.mock.Mock'>: it's not the same object as unittest.mock.Mock

../../../../venv/celery/lib/python3.6/site-packages/kombu/serialization.py:350: EncodeError
____________________ test_Request.test_terminate__pool_ref _____________________

wrapper = <class 'kombu.exceptions.EncodeError'>
include = (<class 'Exception'>,)
exclude = (<class 'kombu.exceptions.SerializerNotInstalled'>,)

    @contextmanager
    def _reraise_errors(wrapper,
                        include=(Exception,), exclude=(SerializerNotInstalled,)):
        try:
>           yield

../../../../venv/celery/lib/python3.6/site-packages/kombu/serialization.py:50: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <kombu.serialization.SerializerRegistry object at 0x7fbc7b2ffcc0>
data = {'args': None, 'children': [], 'date_done': datetime.datetime(2018, 8, 15, 16, 37, 27, 706803), 'kwargs': None, ...}
serializer = 'pickle'
```

It's too bad TravisCI truncated the old builds.

I'll revert this, and the one below and let TravisCI re-run so you can see.